src/analysis/raster/qgskde.cpp

```diff
@@ -90,16 +83,35 @@ QgsKernelDensityEstimation::Result QgsKernelDensityEstimation::prepare()
   int rows = qMax( ceil( mBounds.height() / mPixelSize ) + 1, 1.0 );
   int cols = qMax( ceil( mBounds.width() / mPixelSize ) + 1, 1.0 );
 
-  if ( !createEmptyLayer( driver, mBounds, rows, cols ) )
+  // create empty raster and fill it with nodata values
+  QgsRasterFileWriter writer( mOutputFile );
+  writer.setOutputProviderKey( QStringLiteral( "gdal" ) );
+  writer.setOutputFormat( mOutputFormat );
+  mProvider = writer.createOneBandRaster( Qgis::Float32, cols, rows, mBounds, mSource->sourceCrs() );
+  if ( !mProvider )
+  {
     return FileCreationError;
+  }
 
-  // open the raster in GA_Update mode
-  mDatasetH = GDALOpen( mOutputFile.toUtf8().constData(), GA_Update );
-  if ( !mDatasetH )
-    return FileCreationError;
-  mRasterBandH = GDALGetRasterBand( mDatasetH, 1 );
-  if ( !mRasterBandH )
+  if ( !mProvider->setNoDataValue( 1, NO_DATA ) )
+  {
     return FileCreationError;
+  }
+
+  int dataTypeSize = QgsRasterBlock::typeSize( Qgis::Float32 );
+  float *line = new float[ cols ];
```

Comment:
Can this use `std::vector` or `std::unique_ptr` to be less prone to leak memory?

Reply:
Should be fixed now